### ISSUE TITLE
[FEATURE]: Poisson enrichment loss

### ISSUE THREAD

**Issue by matteoaldeghi** [chemprop/chemprop#612]
**Use-cases/examples of this new feature**
Model training on data from DNA-encoded library screening.

**Desired solution/workflow**
Implement the custom loss described in this paper: https://pubs.acs.org/doi/abs/10.1021/acs.jcim.2c00041

**Discussion**
It could be implemented as an additional loss to be selected via `--loss_function` while also expecting two `--target_columns` with the count data.

**Comment by JacksonBurns** [comment]
Hi @matteoaldeghi! We are probably not going to add any new features that aren't already coded into Chemprop v1 - are you still interested in this feature being added if it doesn't make it in until v2, which is still a way out from release?

**Comment by danielvik** [comment]
Any updates on this? I'd be very interested in a stable implementation on this. I can read that you are considering a larger overhaul of the metrics/loss architecture, but it's not clear to me what the status is based on the milestones etc. 

For the time being -- and if it's of use to others e.g. @matteoaldeghi -- I did a hacky implementation of the loss function described in the Lim et al. paper.


Added a new loss function to nn.metrics.py:

`
@LossFunctionRegistry.register("nlogprob-enrichment")
class NLogProbEnrichment(ChempropMetric):

    def __init__(
        self,
        task_weights: ArrayLike = 1.0,
        n1: int = 1,
        n2: int = 1,
        method: str = "sqrt",
        zscale: float = 1.0,

    ):
        super().__init__(task_weights)
        self.n1 = n1
        self.n2 = n2
        self.method = method
        self.zscale = zscale



    def _calc_unreduced_loss(
        self,
        preds: Tensor,
        targets: Tensor,
        mask: Tensor,
        weights: Tensor,
        *args
    ) -> Tensor:
        # Assuming `preds` are enrichment values R
        R = preds.squeeze()
        #print('R:', R)


        k1 = targets[:, 0]
        k2 = targets[:, 1]

        zstat = get_zstats(R, k1, k2, self.n1, self.n2, method=self.method)
        zstat = torch.clamp(zstat / self.zscale, -5, 5)
        zstat = torch.abs(zstat)
        sf = 1 - torch.erf(zstat / np.sqrt(2))
        loss = -torch.log(sf)
        return loss.unsqueeze(1)

    def extra_repr(self):
        return f"n1={self.n1}, n2={self.n2}, method='{self.method}', zscale={self.zscale}"

`


Modified RegressionFFN in nn.predictors.py to return only positive R values
`@PredictorRegistry.register("regression")
class RegressionFFN(_FFNPredictorBase):
    n_targets = 1
    _T_default_criterion = MSE
    _T_default_metric = MSE

    def forward(self, Z: Tensor) -> Tensor:
        # return self.output_transform(self.ffn(Z))
        output = self.ffn(Z)
        output = F.softplus(output)  # ensures R > 0
        return self.output_transform(output)

    train_step = forward
`

Using the hyperopting example as base (https://github.com/chemprop/chemprop/blob/main/examples/hpopting.ipynb) I add k1 and k2 to the target_columns, and modified the train_model function by adding n1 and n2, and removing the output_transform.
`def train_model(config, 
                train_dset, 
                val_dset, 
                num_workers, 
                #scaler
                ):

    # config is a dictionary containing hyperparameters used for the trial
    depth = int(config["depth"])
    ffn_hidden_dim = int(config["ffn_hidden_dim"])
    ffn_num_layers = int(config["ffn_num_layers"])
    message_hidden_dim = int(config["message_hidden_dim"])

    train_loader = data.build_dataloader(train_dset, num_workers=num_workers, shuffle=True)
    val_loader = data.build_dataloader(val_dset, num_workers=num_workers, shuffle=False)

    mp = nn.BondMessagePassing(d_h=message_hidden_dim, depth=depth)
    agg = nn.MeanAggregation()
    #output_transform = nn.UnscaleTransform.from_standard_scaler(scaler)
    ffn = predictors_modified.RegressionFFN(#output_transform=output_transform, 
                           input_dim=message_hidden_dim, 
                           hidden_dim=ffn_hidden_dim, 
                           n_layers=ffn_num_layers,
                           criterion=metrics_modified.NLogProbEnrichment(n1 = total_counts_poi,
                                                                         n2 = total_counts_counter,
                                                                         method="sqrt", 
                                                                         zscale=1.0,
                                                                         ))
    batch_norm = True
    metric_list = [
        #nn.metrics.RMSE(), 
        #nn.metrics.MAE(), 
        metrics_modified.NLogProbEnrichment(n1=total_counts_poi, n2=total_counts_counter)
    ]
    model = models.MPNN(mp, agg, ffn, batch_norm, metric_list)

    trainer = pl.Trainer(
        accelerator="auto",
        devices=1,
        gradient_clip_val=1.0, 
        max_epochs=20, # number of epochs to train for
        # below are needed for Ray and Lightning integration
        strategy=RayDDPStrategy(),
        callbacks=[RayTrainReportCallback()],
        plugins=[RayLightningEnvironment()],
    )

    trainer = prepare_trainer(trainer)
    trainer.fit(model, train_loader, val_loader)`


I could do a PR, but I dont know if you are planning on changing the metrics and loss functions architecture. 

Thanks for making an awesome module!
Cheers
